Objects/bytesobject.c

```diff
@@ -2810,82 +2811,44 @@ _PyBytes_FromBuffer(PyObject *x)
 }
 
 static PyObject*
-_PyBytes_FromList(PyObject *x)
+_PyBytes_FromSequence(PyObject *x)
 {
-    Py_ssize_t i, size = PyList_GET_SIZE(x);
-    Py_ssize_t value;
-    char *str;
-    PyObject *item;
-    _PyBytesWriter writer;
-
-    _PyBytesWriter_Init(&writer);
-    str = _PyBytesWriter_Alloc(&writer, size);
-    if (str == NULL)
+    Py_ssize_t size = PySequence_Fast_GET_SIZE(x);
+    PyObject *bytes = _PyBytes_FromSize(size, 0);
+    if (bytes == NULL) {
         return NULL;
-    writer.overallocate = 1;
-    size = writer.allocated;
-
-    for (i = 0; i < PyList_GET_SIZE(x); i++) {
-        item = PyList_GET_ITEM(x, i);
-        Py_INCREF(item);
-        value = PyNumber_AsSsize_t(item, NULL);
-        Py_DECREF(item);
-        if (value == -1 && PyErr_Occurred())
-            goto error;
-
-        if (value < 0 || value >= 256) {
-            PyErr_SetString(PyExc_ValueError,
-                            "bytes must be in range(0, 256)");
-            goto error;
-        }
-
-        if (i >= size) {
-            str = _PyBytesWriter_Resize(&writer, str, size+1);
-            if (str == NULL)
-                return NULL;
-            size = writer.allocated;
-        }
-        *str++ = (char) value;
     }
-    return _PyBytesWriter_Finish(&writer, str);
-
-  error:
-    _PyBytesWriter_Dealloc(&writer);
-    return NULL;
-}
-
-static PyObject*
-_PyBytes_FromTuple(PyObject *x)
-{
-    PyObject *bytes;
-    Py_ssize_t i, size = PyTuple_GET_SIZE(x);
-    Py_ssize_t value;
-    char *str;
-    PyObject *item;
-
-    bytes = PyBytes_FromStringAndSize(NULL, size);
-    if (bytes == NULL)
-        return NULL;
-    str = ((PyBytesObject *)bytes)->ob_sval;
-
-    for (i = 0; i < size; i++) {
-        item = PyTuple_GET_ITEM(x, i);
-        value = PyNumber_AsSsize_t(item, NULL);
-        if (value == -1 && PyErr_Occurred())
+    char *str = PyBytes_AS_STRING(bytes);
+    PyObject *const *items = PySequence_Fast_ITEMS(x);
+    Py_BEGIN_CRITICAL_SECTION_SEQUENCE_FAST(x);
+    for (Py_ssize_t i = 0; i < size; i++) {
+        if (!PyLong_Check(items[i])) {
+            Py_DECREF(bytes);
+            /* Py_None as a fallback sentinel to the slow path */
+            bytes = Py_None;
+	    goto done;
+        }
+        int overflow;
+        long value = PyLong_AsLongAndOverflow(items[i], &overflow);
+        if (value == -1 && PyErr_Occurred()) {
             goto error;
-
+        }
         if (value < 0 || value >= 256) {
+            /* this includes an overflow in converting to C long */
             PyErr_SetString(PyExc_ValueError,
                             "bytes must be in range(0, 256)");
             goto error;
         }
         *str++ = (char) value;
     }
-    return bytes;
-
+    goto done;
   error:
     Py_DECREF(bytes);
-    return NULL;
+    bytes = NULL;
+  done:
+    /* both success and failure need to end the critical section */
+    Py_END_CRITICAL_SECTION_SEQUENCE_FAST();
```

Comment:
> label at end of compound statement

If you expand Py_END_CRITICAL_SECTION_SEQUENCE_FAST(), it will be a simple `}` I would advise doing something like that to make the compiler happy: (I think it should work, haven't checked)

```c
done:
	;
```

Reply:
I did not get such a complaint with gcc in Linux and was scratching my head when I saw this error in CI. Makes sense. Thanks!